server/src/main/java/io/crate/execution/dml/ObjectIndexer.java

```diff
@@ -94,35 +99,58 @@ public void indexValue(@NotNull Map<String, Object> value, IndexDocumentBuilder
         for (var entry : children.entrySet()) {
             String innerName = entry.getKey();
             Child child = entry.getValue();
-            Object innerValue = null;
             if (value.containsKey(innerName) == false) {
-                innerValue = docBuilder.getSyntheticValue(child.ident());
-            } else {
-                innerValue = value.get(innerName);
+                var synth = docBuilder.getSyntheticValue(child.ident());
+                if (synth != null) {
+                    // directly modify the map so that containing types will see the value
+                    // if they need to write stored fields
+                    value.put(innerName, synth);
+                }
             }
+            var innerValue = value.get(innerName);
             docBuilder.checkColumnConstraint(child.ident(), innerValue);
             if (innerValue == null) {
                 continue;
             }
             var valueIndexer = child.indexer;
             // valueIndexer is null for partitioned columns
             if (valueIndexer != null) {
+                innerValue = child.reference.valueType().sanitizeValue(innerValue);
                 docBuilder.translogWriter().writeFieldName(child.reference.storageIdentLeafName());
-                valueIndexer.indexValue(
-                    child.reference.valueType().sanitizeValue(innerValue),
-                    docBuilder
-                );
+                valueIndexer.indexValue(innerValue, docBuilder);
+                value.put(innerName, innerValue);
             }
         }
+        Map<String, Object> columnsToStore = new HashMap<>();
         value.forEach((k, v) -> {
             if (children.containsKey(k) == false) {
                 translogWriter.writeFieldName(this.unknownColumnPrefix + k);
                 translogWriter.writeValue(v);
+                columnsToStore.put(k, v);
+            }
+            if (v == null) {
+                columnsToStore.put(k, null);
             }
         });
+        if (docBuilder.maybeAddStoredField()) {
+            if (columnsToStore.isEmpty() == false) {
+                docBuilder.addField(new StoredField(ref.storageIdentLeafName(), toBytes(columnsToStore).toBytesRef()));
+            } else if (value.isEmpty()) {
+                docBuilder.addField(new StoredField(ref.storageIdentLeafName(), toBytes(value).toBytesRef()));
+            }
+        }
```

Comment:
Maybe also add a short comment here explaining why the object needs to store selected key/values separately.